orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/ServerGroupCacheForceRefreshTask.groovy

```diff
@@ -154,26 +159,57 @@ class ServerGroupCacheForceRefreshTask extends AbstractCloudProviderAwareTask im
                                                   String cloudProvider,
                                                   StageData stageData,
                                                   Long startTime) {
+
     def pendingForceCacheUpdates = cacheStatusService.pendingForceCacheUpdates(cloudProvider, REFRESH_TYPE)
+    log.debug("Force cache refresh clouddriver response was: ${pendingForceCacheUpdates}.")
     boolean isRecent = (startTime != null) ? pendingForceCacheUpdates.find { it.cacheTime >= startTime } : false
 
+    log.debug("Force cache refresh isRecent = ${isRecent}")
+
     boolean finishedProcessing = true
     stageData.deployServerGroups.each { String region, Set<String> serverGroups ->
-      def makeModel = { serverGroup ->
-        [serverGroup: serverGroup, region: region, account: account]
-      }
+      def makeModel = { serverGroup -> [serverGroup: serverGroup, region: region, account: account] }
 
       def processedServerGroups = serverGroups.findAll { String serverGroup ->
         def model = makeModel(serverGroup)
 
+        def forceCacheUpdate = pendingForceCacheUpdates.find {
+          (it.details as Map<String, String>).intersect(model) == model
+        }
+
+        if (!stageData.processedServerGroups.contains(model) && forceCacheUpdate == null) {
+          log.debug("Force cache refresh response narrowed to ${model} was null, and this group hasn't been processed yet!")
+        }
+
         if (stageData.processedServerGroups.contains(model)) {
           // this server group has already been processed
+          log.debug("Force cache refresh the model ${model} has been already processed, returning true.")
           return true
         }
 
+        if (!forceCacheUpdate) {
+          // there is no pending cache update, force it again in the event that it was missed
+          stageData.removeRefreshedServerGroup(model.serverGroup, model.region, model.account)
+          log.warn("Unable to find pending cache refresh request (model: ${model}), forcing a new cache refresh.")
+
+          try {
+            log.info("Force immediate cache refresh POST to clouddriver with model: ${model}")
+            def response = cacheService.forceCacheUpdate(cloudProvider, REFRESH_TYPE, model)
+            if (response.status == HttpURLConnection.HTTP_OK) {
+              // cache update was applied immediately, no need to poll for completion
+              log.info("Processed force cache refresh request immediately (model: ${model})")
+              return true
+            }
+            stageData.refreshedServerGroups << model
+          } catch (e) {
+            stageData.errors << e.message
+          }
+          return false
+        }
+
         if (!isRecent) {
           // replication lag -- there are no pending force cache refreshes newer than this particular stage ... retry in 10s
-          log.debug(
+          log.warn(
```

Comment:
I'm not sure I agree with this escalation of log level, nor the one below at the end of the function. Can you explain why you believe these should be louder?

Reply:
Probably because it should not really happen? /shrug If you feel like it is regular and typical thing, then yeah, we should leave it at the debug level.